[lldb][LoongArch64] Add support for LoongArch64 in elf-core for lldb

lawn123 opened the pull request:
When using the lldb command 'target create --core' on the LoongArch64 architecture, this part of the code is required.

lawn123 committed:
[LoongArch64]: Add support for LoongArch64 in elf-core for lldb

lawn123 commented:
@shushanhf

shushanhf commented:
> @shushanhf

@SixWeining

SixWeining commented on lldb/source/Plugins/Process/elf-core/RegisterContextPOSIXCore_loongarch64.cpp:
```diff
@@ -0,0 +1,87 @@
+//===-- RegisterContextPOSIXCore_loongarch64.cpp
+//------------------------------===//
```
format

SixWeining commented on lldb/source/Plugins/Process/elf-core/RegisterContextPOSIXCore_loongarch64.h:
```diff
@@ -0,0 +1,62 @@
+//===-- RegisterContextPOSIXCore_loongarch64.h ----------------------*- C++
+//-*-===//
```
ditto

SixWeining reviewed (commented):
@wangleiat may take a look.

SixWeining commented:
> When using the lldb command 'target create -- core' on the LoongArch64 architecture, this part of the code is required.

Do you mean `--core` but not `-- core`?

lawn123 commented:
> > When using the lldb command 'target create -- core' on the LoongArch64 architecture, this part of the code is required.
> 
> Do you mean `--core` but not `-- core`?

Yes. use ' target create --core' to load dump files

DavidSpickett commented:
I don't see any tests yet, I think `lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py` will be the place for those. There is a template program file you can use in that folder.

If you need more functionality to get to the point you can add tests, please add that to this PR or explain your plan for getting to a point where testing is possible (I'm not aware of the current status of LoongArch in LLDB).

SixWeining commented on lldb/source/Plugins/Process/elf-core/RegisterContextPOSIXCore_loongarch64.h:
```diff
@@ -0,0 +1,62 @@
+  bool WriteFPR() override;
+
+private:
+  lldb::DataBufferSP m_gpr_buffer;
```
Seems we only need to define them as local variables before using it. No need to be class-member-variables.

SixWeining commented:
Seems we the same `GetRegisterSetCount` misprint issue as #93297.

lawn123 committed:
Fix formatting and add two testcases in TestLinuxCore.py

lawn123 commented:
> I don't see any tests yet, I think `lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py` will be the place for those. There is a template program file you can use in that folder.
@DavidSpickett @SixWeining  I have added two test cases to `TestLinuxCore. py` and modified the format as well as the issue with `GetRegisterSetCount`

SixWeining commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
+        fpr_values["fcc3"] = "0x01"
+        fpr_values["fcc4"] = "0x01"
+        fpr_values["fcc5"] = "0x01"
+        fpr_values["fcc6"] = "0x0"
```
Ditto.

SixWeining commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
+        fpr_values["f30"] = "0xffffffffffffffff"
+        fpr_values["f31"] = "0xffffffffffffffff"
+        fpr_values["fcc0"] = "0x01"
+        fpr_values["fcc1"] = "0x0"
```
Should it be `0x00`? Otherwise `0x01` can also match the expect.

SixWeining commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
+        fpr_values["fcc5"] = "0x01"
+        fpr_values["fcc6"] = "0x0"
+        fpr_values["fcc7"] = "0x01"
+        fpr_values["fcsr"] = "0x0"
```
0x00000000

lawn123 replied:
> 0x00000000

Other formatting issues have been corrected. Here, only 0x01 and 0x00 are allowed. Otherwise, the test will not pass

SixWeining commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
+        values["r11"] = "0x00000000000000dd"
+        values["r12"] = "0x0"
+        values["r13"] = "0x000000000000002f"
+        values["r14"] = "0x0"
```
Should we use full encoding (i.e. 8bytes)?

SixWeining commented on lldb/source/Plugins/Process/elf-core/RegisterContextPOSIXCore_loongarch64.cpp:
```diff
@@ -0,0 +1,84 @@
+//===-- RegisterContextPOSIXCore_loongarch64.cpp ------------------------------===//
```
Generally one line has 80 characters at most.

SixWeining commented:
> I'm not aware of the current status of LoongArch in LLDB

@DavidSpickett General debugging commands can be used on LoongArch with some known issues to be addressed. Here is a summary #109394.

For this PR, I have checked locally. I can confirm core dump file could be debugged with this change.

lawn123 committed:
Fix format

DavidSpickett commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
 
+    @skipIfLLVMTargetMissing("LoongArch")
+    def test_loongarch64_regs(self):
+        # check registers using 64 bit LoongArch64 core file containing GP-registers only
```
Maybe this is a terminology difference. To me GP register would mean "general purpose" and not include floating point registers. But perhaps you always have an FPU so they are kinda general in that they always exist.

I'd just change it to "GP and FP registers" to prevent any confusion.

lawn123 replied:
There is confusion, I will revise it later.

lawn123 replied:
There is confusion, I will revise it later.

DavidSpickett commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
             substrs=["registers were unavailable"],
         )
 
+    @skipIfLLVMTargetMissing("LoongArch")
```
Is LoongArch one LLVM target or do you need to specifically have LoongArch64 for this?

DavidSpickett commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
+        fpr_values["fcc5"] = "0x01"
+        fpr_values["fcc6"] = "0x00"
+        fpr_values["fcc7"] = "0x01"
+        fpr_values["fcsr"] = "0x00"
```
Are these 8 bit registers?

SixWeining replied:
Not like fcc registers, fcsr is 32bit.
```
(lldb) register read fcsr
    fcsr = 0x00000000
```

DavidSpickett commented:
> @DavidSpickett General debugging commands can be used on LoongArch with some known issues to be addressed. Here is a summary https://github.com/llvm/llvm-project/issues/109394.

For RISC-V I have been using https://github.com/llvm/llvm-project/issues/55383 as the tracking issue for overall support. If you want to, you could open a similar issue and write out what currently works, and I'll link to it from the [main lldb page](https://lldb.llvm.org/#platform-support). Not a requirement, just if you have the time.

DavidSpickett commented on lldb/test/API/functionalities/postmortem/elf-core/linux-loongarch64.core:
You may be able to reduce the file by changing the `/proc/pid/coredump_filter` value: https://man7.org/linux/man-pages/man5/core.5.html

I can't say I understand them all but usually what I do is cut things until the test fails. I've been able to get Arm files down to around 20kb doing things, but you don't have to take that as a target.

lawn123 replied:
> You may be able to reduce the file by changing the `/proc/pid/coredump_filter` value: https://man7.org/linux/man-pages/man5/core.5.html

I reduced the core file size to 36kb.

SixWeining commented:
> For RISC-V I have been using #55383 as the tracking issue for overall support. If you want to, you could open a similar issue and write out what currently works, and I'll link to it from the [main lldb page](https://lldb.llvm.org/#platform-support). Not a requirement, just if you have the time.

Thanks. I just filed #112693.

lawn123 committed:
Fix format and reduce the size of the core file

DavidSpickett commented:
Not sure what's going on with the register formatting. It might be that because we do a substring comparison, `foo = 0x00` is found in `foo = 0x00000000`.

So I would recommend loading the core file manually and just copying whatever `register read` shows. If the formatting is still off, potentially you need to update the register sizes or formatting style, but I don't think you need to do that in this PR. It can be a follow up to do just that.

Other than that I don't have any objections here, seems like @SixWeining is the architecture expert so they can give the final approval.

SixWeining commented on lldb/test/API/functionalities/postmortem/elf-core/TestLinuxCore.py:
```diff
@@ -833,6 +845,107 @@ def test_riscv64_regs_gpr_only(self):
+        fpr_values["fcc5"] = "0x01"
+        fpr_values["fcc6"] = "0x00"
+        fpr_values["fcc7"] = "0x01"
+        fpr_values["fcsr"] = "0x0000"
```
(lldb) register read fcsr 
    fcsr = 0x00000000

SixWeining reviewed (commented):
Looks good to me once the `code_formatter` CI fail and fcsr format issues are addressed.

lawn123 committed:
Fix CI fail and fcsr format issues

SixWeining reviewed (approved):
LGTM and thanks!
